### DISCUSSION TITLE
build123d and cq_warehouse

### DISCUSSION THREAD

**Discussion by etherealmccoy** [gumyr/build123d#199] [Q&A]
Hi gumyr. Thank you for the great work on cadquery, cq_warehouse, and build123d. 

@anyone : I am trying to switch a small project from cadquery to build123d. In cadquery, I was using IsoThread from cq_warehouse. It is not clear to me how to use cq_warehouse objects in build123d. I know that underneath it all, we're working with opencascade, so I would think it is possible, but it is not yet apparent to me. Would appreciate any help you can provide in the meantime!

**Comment by Jojain** [comment]
Since it has not been ported yet and CQ and b123d are not compatible on the direct Api level if you want to use the cq_warehouse object you will need to created them in CQ export them as step or BREP and import them back in b123d

That seems to be the simplest, if you don't want to do import/export you can translate a shape from CQ to b123d by going down to the OCP level

**Comment by etherealmccoy** [comment]
Ok. Thank you.Sent from my iPhoneOn Apr 4, 2023, at 1:36 AM, Jojain ***@***.***> wrote:﻿
Since it has not been ported yet and CQ and b123d are not compatible on the direct Api level if you want to use the cq_warehouse object you will need to created them in CQ export them as step or BREP and import them back in b123d
That seems to be the simplest, if you don't want to do import/export you can translate a shape from CQ to b123d by going down to the OCP level

**Comment by gumyr** [comment]
As @Jojain pointed out, I haven't had time to port cq_warehouse over to build123d yet although I plan to do this after the first release (hopefully in the next couple months). 

As they are both based on OCP, you should be able to create an object in cq then use the `.wrapped` (OCP) value to create a new b3d object like:
```py
import cadquery as cq
import build123d as bd

my_cq_solid = ...
my_b3d_solid = bd.Solid(my_cq_solid.wrapped)
```

**Comment by etherealmccoy** [comment]
Very helpful. Thank you, gumyr. This software (both cadquery and build123d) is a big step up from openscad.Sent from my iPhoneOn Apr 4, 2023, at 10:46 AM, gumyr ***@***.***> wrote:﻿
As @Jojain pointed out, I haven't had time to port cq_warehouse over to build123d yet although I plan to do this after the first release (hopefully in the next couple months).
As they are both based on OCP, you should be able to create an object in cq then use the .wrapped (OCP) value to create a new b3d object like:
import cadquery as cq
import build123d as bd

my_cq_solid = ...
my_b3d_solid = bd.Solid(my_cq_solid.wrapped)

**Comment by etherealmccoy** [comment]
By the way‚ would love to contribute, but no time. Is there a way to donate $?Sent from my iPhoneOn Apr 4, 2023, at 7:12 PM, Ethereal McCoy ***@***.***> wrote:﻿Very helpful. Thank you, gumyr. This software (both cadquery and build123d) is a big step up from openscad.Sent from my iPhoneOn Apr 4, 2023, at 10:46 AM, gumyr ***@***.***> wrote:﻿
As @Jojain pointed out, I haven't had time to port cq_warehouse over to build123d yet although I plan to do this after the first release (hopefully in the next couple months).
As they are both based on OCP, you should be able to create an object in cq then use the .wrapped (OCP) value to create a new b3d object like:
import cadquery as cq
import build123d as bd

my_cq_solid = ...
my_b3d_solid = bd.Solid(my_cq_solid.wrapped)

**Reply by gumyr** [reply]
It took a while to get the approvals from GitHub but there is now a `Sponsor this project` section on the GitHub repo main page if you'd like to make a donation.
![image](https://user-images.githubusercontent.com/85638518/231207520-9d1b0d14-441c-44f0-a9fb-61d8d009da20.png)

**Comment by gumyr** [comment]
> Hi gumyr. Thank you for the great work on cadquery, cq_warehouse, and build123d.
> 
> @anyone : I am trying to switch a small project from cadquery to build123d. In cadquery, I was using IsoThread from cq_warehouse. It is not clear to me how to use cq_warehouse objects in build123d. I know that underneath it all, we're working with opencascade, so I would think it is possible, but it is not yet apparent to me. Would appreciate any help you can provide in the meantime!

There is now a bd_warehouse project that has threads: https://bd-warehouse.readthedocs.io/en/latest/thread.html and https://github.com/gumyr/bd_warehouse. The API is very close to identical to the cq_warehouse version; however, the implementation has been ported to build123d and optimized.  Please let me know if you have any questions about it.
